src/useRouteQueryParam/useRouteQueryParam.ts

```diff
@@ -1,55 +1,46 @@
 /* eslint-disable no-redeclare */
-import { Ref, ref, watch } from 'vue'
-import { onBeforeRouteLeave, RouteLocationNormalized, useRoute, useRouter } from 'vue-router'
-import { isSame } from '@/utilities/variables'
-
-export function useRouteQueryParam(param: string): Ref<string | string[]>
-export function useRouteQueryParam(param: string, defaultValue: string): Ref<string>
-export function useRouteQueryParam(param: string, defaultValue: string[]): Ref<string[]>
-export function useRouteQueryParam(param: string, defaultValue: string | string[] = ''): Ref<string | string[]> {
-  const router = useRouter()
-  const route = useRoute()
-  const initialValue = matchValueType(defaultValue, getRouteQueryParam(route, param) ?? defaultValue)
-  const valueRef = ref<string | string[]>(initialValue)
-
-  const unwatchValue = watch(valueRef, (newValue, oldValue) => {
-    if (isSame(newValue, oldValue)) {
-      return
-    }
-
-    router.push({ query: { ...route.query, [param]: valueRef.value } })
-  })
+import { computed, Ref } from 'vue'
+import { NoInfer } from '@/types/generics'
+import { MaybeArray } from '@/types/maybe'
+import { useRouteQuery } from '@/useRouteQuery/useRouteQuery'
+import { isRouteParamClass, RouteParamClass } from '@/useRouteQueryParam/formats/RouteParam'
+import { StringRouteParam } from '@/useRouteQueryParam/formats/StringRouteParam'
+
+function isDefaultValue<T>(value: T | RouteParamClass): value is T {
+  return !isRouteParamClass(value)
+}
 
-  const unwatchRoute = watch(route, (newRoute, oldRoute) => {
-    const newValue = getRouteQueryParam(newRoute, param) ?? defaultValue
-    const oldValue = getRouteQueryParam(oldRoute, param) ?? defaultValue
-    const matched = matchValueType(oldValue, newValue)
+export function useRouteQueryParam(key: string, defaultValue?: string): Ref<string>
```

Comment:
I know this composition has gone through a few iterations, I just want to say that the way it turned out is quite easy to understand and just generally fantastic

Reply:
❤️